network/src/dms.rs

```diff
@@ -1,65 +1,149 @@
+use super::Storage;
 use super::*;
-use async_trait::async_trait;
-use simperby_common::{crypto::*, BlockHeight};
+use anyhow::anyhow;
+use serde_tc::http::*;
+use serde_tc::{serde_tc_full, StubCall};
+use simperby_common::*;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+#[derive(Debug, Clone, Serialize)]
+pub struct Message {
+    data: String,
+    signature: TypedSignature<String>,
+}
+
+impl Message {
+    pub fn new(data: String, signature: TypedSignature<String>) -> Result<Self, CryptoError> {
+        signature.verify(&data)?;
+        Ok(Self { data, signature })
+    }
+
+    pub fn data(&self) -> &str {
+        &self.data
+    }
+
+    pub fn signature(&self) -> &TypedSignature<String> {
+        &self.signature
+    }
+}
+
+impl ToHash256 for Message {
+    fn to_hash256(&self) -> Hash256 {
+        Hash256::hash(serde_json::to_vec(self).unwrap())
+    }
+}
+
+/// A message before verification.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+struct RawMessage {
+    pub data: String,
+    pub signature: TypedSignature<String>,
+}
+
+impl RawMessage {
+    #[allow(dead_code)]
+    pub fn into_message(self) -> anyhow::Result<Message> {
+        Message::new(self.data, self.signature).map_err(|e| anyhow!(e))
+    }
+
+    #[allow(dead_code)]
+    pub fn from_message(message: Message) -> Self {
+        RawMessage {
+            data: message.data().to_owned(),
+            signature: message.signature().to_owned(),
+        }
+    }
+}
 
 /// The interface that will be wrapped into an HTTP RPC server for the peers.
-#[async_trait]
-pub trait DistributedMessageSetRpcInterface {
-    /// Returns the messages except `knowns`. If the height is different, it returns `Err(height)`.
+#[serde_tc_full]
+trait DistributedMessageSetRpcInterface: Send + Sync + 'static {
+    /// Returns the messages except `knowns`.
     async fn get_message(
         &self,
         height: BlockHeight,
         knowns: Vec<Hash256>,
-    ) -> Result<Vec<Message>, BlockHeight>;
+    ) -> Result<Vec<RawMessage>, String>;
 }
 
-pub struct DistributedMessageSetImpl {}
+/// A **cumulative** set that is shared in the p2p network, backed by the local file system.
+///
+/// One of the notable characteristics of blockchain is that it is based on heights;
+/// The key idea here is that we retain an instance (both in memory or on disk)
+/// of `DistributedMessageSet` only for a specific height,
+/// and discard if the height progresses, creating a new and empty one again.
+///
+/// For every methods,
```

Comment:
```suggestion
/// For every method,
```

Reply:
this comment will be replaced in total in the next PR. Let's keep it for now.